Replaced greenDAO with androidx.room

JaniruTEC opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced Room library for enhanced database management.
  - Added new tests for database creation, access, and migrations.
  - Implemented automatic and manual database migrations for ~improved~ data integrity.
  - Added new data access objects (DAOs) for Cloud, Update Check, and Vault entities.

- **Enhancements**
  - ~Improved database operation efficiency by utilizing Kotlin's type alias and constructor-based entity creation.~
  - ~Enhanced database schema with the addition of entities for Cloud, Update Check, and Vault.~
  - Streamlined database access in repository implementations with dependency injection.

- **Bug Fixes**
  - Fixed issues related to database migrations and entity mapping for a more stable user experience.

- **Refactor**
  - Transitioned from GreenDao to Room for database operations, resulting in cleaner and more maintainable code.
  - Refined database migration logic to use Room's built-in migration features.

- **Documentation**
  - Updated internal comments to reflect changes in database handling and migration processes.

- **Chores**
  - Updated build configurations to include necessary Room library dependencies.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

JaniruTEC committed:
Started Dao Migration

Replaced greenDAO dependency with androidx.room
Introduced database version 13

JaniruTEC committed:
Simplified database schema

Introduced database version 14

JaniruTEC committed:
Merge branch 'develop' into feature/dao-migration

JaniruTEC committed:
Removed unnecessary historic method "clearCache"

The method was a delegate to the "DaoSession#clear" method provided by greenDAO.
Analysis of the generated code indicates that room doesn't cache results internally. Therefore clearing the cache to evict stale entries is unnecessary.

See: https://web.archive.org/web/20230806043911/https://greenrobot.org/greendao/documentation/sessions/

JaniruTEC committed:
Added logging/comments to database configuration

JaniruTEC committed:
Added logging to legacy/manual migrations

JaniruTEC committed:
Reorganized migrations

JaniruTEC committed:
Refactored injection of Migration12To13

JaniruTEC committed:
Added logging to auto migrations

JaniruTEC committed:
Merge branch 'develop' into feature/dao-migration

JaniruTEC committed:
Reformatted "org.cryptomator.data.db" package

[ci skip]

JaniruTEC committed:
Externalized database version for use in other locations

JaniruTEC committed:
Started fixing UpgradeDatabaseTest

JaniruTEC committed:
Reformatted UpgradeDatabaseTest

JaniruTEC committed:
Fixed formatting introduced by merge commit 4cce1eb3

JaniruTEC committed:
Fixed regression in Sql: Caller is now allowed to omit columns to receive all

JaniruTEC committed:
Added and configured "room-testing"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cryptomator/android?pullRequest=506) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cryptomator/android?pullRequest=506) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cryptomator/android?pullRequest=506) it.</sub>

JaniruTEC committed:
Changed configuration to use template db for creation of app db

Without this change room will create the app db itself on installation without running the migrations.

JaniruTEC committed:
Changed timing of migrations

JaniruTEC committed:
Merge branch 'develop' into feature/dao-migration

JaniruTEC committed:
Added first tests for migrations involving room using "room-testing"

JaniruTEC committed:
Updated tests to use template db

JaniruTEC committed:
Refactored database to be initialized on demand

JaniruTEC committed:
Refactored CloudRepositoryImpl to use CloudDao directly

JaniruTEC committed:
Replaced asset-based template with runtime-generated file

JaniruTEC committed:
Wrapped runtime-generated template with lazy stream

JaniruTEC committed:
Added template generation to dagger graph

JaniruTEC committed:
Cleaned up DatabaseModule

Used dedicated logger
Added "DbInternal" qualifier to not pollute the dagger graph
Reformatted file

JaniruTEC committed:
Added warning if database is built on main thread

JaniruTEC committed:
Updated tests for migrations to use new template db mechanism

JaniruTEC committed:
Added exception and performed minor cleanup

JaniruTEC committed:
Extracted TemplateDatabaseContext from object

JaniruTEC committed:
Added tests and changed contract of TemplateDatabaseContext

JaniruTEC committed:
Separated room schemas from androidTest assets

JaniruTEC committed:
Changed action for foreign key in "VaultEntity" to "RESTRICT"

JaniruTEC committed:
Cleaned up and improved implementations in "Sql.java"

JaniruTEC committed:
Fixed wrong test

JaniruTEC committed:
Updated "Sql.java" to offer new id column

JaniruTEC commented:
@coderabbitai review

JaniruTEC commented on data/src/main/java/org/cryptomator/data/db/migrations/legacy/Upgrade11To12.kt:
```diff
@@ -1,15 +1,16 @@
 @Singleton
-internal class Upgrade11To12 @Inject constructor(private val sharedPreferencesHandler: SharedPreferencesHandler) : DatabaseUpgrade(11, 12) {
+internal class Upgrade11To12 @Inject constructor(private val sharedPreferencesHandler: SharedPreferencesHandler) : DatabaseMigration(11, 12) {
 
-	override fun internalApplyTo(db: Database, origin: Int) {
+	override fun migrateInternal(db: SupportSQLiteDatabase) {
```
Out of scope of this PR.

JaniruTEC replied:
Please name the issue "Post-merge cleanup/tasks after greenDAO-Room migration (#506)".

Please add your suggestion as part of a task list.
Also please add "`getColumnIndex` warnings" as task as well.

JaniruTEC committed:
Performed minor cleanup

Removed unnecessary comment/constructor
Corrected error message
Replaced unreachable log message with sanity check

JaniruTEC committed:
Enabled foreign key constraints for migrations

JaniruTEC committed:
Added additional cleanup to UpgradeDatabaseTest

JaniruTEC committed:
Cleaned up Entities and Entity mapping

JaniruTEC committed:
Added explicit orders to VaultEntity at v13

The file was generated by applying the following patch onto https://github.com/cryptomator/android/blob/d9611360b20b2aa722bae802c504d716f4021e13/data/src/main/java/org/cryptomator/data/db/entities/VaultEntity.kt

```diff --git a/data/src/main/java/org/cryptomator/data/db/entities/VaultEntity.kt b/data/src/main/java/org/cryptomator/data/db/entities/VaultEntity.kt
   index e99b4ea..202f20e 100644
   --- a/data/src/main/java/org/cryptomator/data/db/entities/VaultEntity.kt
   +++ b/data/src/main/java/org/cryptomator/data/db/entities/VaultEntity.kt
   @@ -4,11 +4,12 @@ import androidx.room.ColumnInfo
    import androidx.room.Entity
    import androidx.room.ForeignKey
    import androidx.room.Index
   +import androidx.room.Index.Order
    import androidx.room.PrimaryKey

    @Entity(
    	tableName = "VAULT_ENTITY", //
   -	indices = [Index(name = "IDX_VAULT_ENTITY_FOLDER_PATH_FOLDER_CLOUD_ID", value = ["FOLDER_PATH", "FOLDER_CLOUD_ID"], unique = true)], //
   +	indices = [Index(name = "IDX_VAULT_ENTITY_FOLDER_PATH_FOLDER_CLOUD_ID", value = ["FOLDER_PATH", "FOLDER_CLOUD_ID"], orders = [Order.ASC, Order.ASC], unique = true)], //
    	foreignKeys = [ForeignKey(CloudEntity::class, ["_id"], ["FOLDER_CLOUD_ID"], onDelete = ForeignKey.SET_NULL)]
    )
    data class VaultEntity constructor(
```

JaniruTEC committed:
Added explicit orders to VaultEntity at current version

JaniruTEC committed:
Added tests to handle mismatch between db and schema

JaniruTEC committed:
Merge branch 'develop' into feature/dao-migration

JaniruTEC committed:
Commented "Cursor?.stringify" defined in CryptomatorDatabase.kt

JaniruTEC committed:
Specified journal mode

Room is designed to choose the journal mode itself if none is specified.
greenDAO's journal mode ("TRUNCATE") will stay for now because the journal mode has influence on the behavior caused by issue #529 [1] and because changing the journal mode complicates the migration further.
This can still be changed later.

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added first implementation of "CacheControlledSupportSQLiteDatabase"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "CacheControlledSupportSQLiteDatabase"

Replaced particular implementation of "fix" with call to freshly introduced, user-supplied mapping function (and added "RandomUUIDSQLMappingFunction" as new mapping function with that particular implementation)
Replaced "fixCompile" with direct instantiation of "CacheControlledSupportSQLiteStatement"
Updated static elements to be inner members of "CacheControlledSupportSQLiteDatabase" (and removed now unnecessary properties)
Changed computation logic of "sql" property in "CacheControlledSupportSQLiteQuery"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Narrowed visibility for members of "AOP_SQLiteDatabase" class

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Renamed parts of cache-control

Renamed package "org.cryptomator.data.db.cachecontrol" to "org.cryptomator.data.db.sqlmapping"
Renamed "CacheControlledSupportSQLiteDatabase.kt" to "MappingSupportSQLiteDatabase.kt"
Moved contents to new location

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Renamed class "CacheControlledSupportSQLiteDatabase" and its members

Renamed class "CacheControlledSupportSQLiteDatabase" to "MappingSupportSQLiteDatabase"
Renamed class "CacheControlledSupportSQLiteOpenHelper" to "MappingSupportSQLiteOpenHelper"
Renamed class "CacheControlledSupportSQLiteOpenHelperFactory" to "MappingSupportSQLiteOpenHelperFactory"
Renamed inner class "CacheControlledSupportSQLiteStatement" to "MappingSupportSQLiteStatement"
Renamed inner class "CacheControlledSupportSQLiteQuery" to "MappingSupportSQLiteQuery"
Renamed method "fix" to "map" in "MappingSupportSQLiteDatabase"
Renamed method "fixWhereClause" to "mapWhereClause" in "MappingSupportSQLiteDatabase"
Renamed property "identifiers" to "mappedQueries" in "MappingSupportSQLiteQuery"
Renamed extension method "SupportSQLiteOpenHelper.Factory.asCacheControlled" to "SupportSQLiteOpenHelper.Factory.asMapped" in "MappingSupportSQLiteDatabase.kt"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Reorganized cache-control to improve abstraction

Moved "RandomUUIDSQLMappingFunction" from "MappingSupportSQLiteDatabase.kt" to new file "SQLiteCacheControl.kt"
Added extension function "SupportSQLiteOpenHelper.Factory.asCacheControlled" in "SQLiteCacheControl.kt" as wrapper for "asMapped" with "RandomUUIDSQLMappingFunction" as mapping function
Removed "RandomUUIDSQLMappingFunction" as default mapping function from extension function "SupportSQLiteOpenHelper.Factory.asMapped" in "MappingSupportSQLiteDatabase.kt"
Replaced call to "asMapped" with call to "asCacheControlled" in "DatabaseModule"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Renamed "RandomUUIDSQLMappingFunction" to "RandomUUIDMapping"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "SQLMappingFunction"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Redefined logic of "MappingSupportSQLiteQuery"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added first tests for "MappingSupportSQLiteDatabase"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Renamed "DataForTestQueryCancelable"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Extracted methods "cartesianProduct" and "toArgumentsStream"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added test for "update"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Specified contract for "SupportSQLiteDatabase.insert"

Also moved compat method for "ContentValues.isEmpty"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added tests for "insert"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Removed erroneous import for "okio.use" in favor of "kotlin.io.use"

Introduced by bfbff577dd629519263998792a5b0f4cce792e26.

JaniruTEC committed:
Added tests for migration to v13

UpgradeDatabaseTest.kt:816 and UpgradeDatabaseTest.kt:895 caused the tests to fail because of issue #529 [1]. Consequently they have been commented out until the issue is resolved.

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Re-enabled tests

See: c13b12ffc8d441f344793f2436506974b145899f

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added utility methods for comparing databases

JaniruTEC committed:
Added "assertCursorEquals"

JaniruTEC committed:
Added tests/snippets to explain/reproduce bug #529 [1]

Copied from the Android-Issue-153521693 project [2]

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529
[2] https://github.com/SailReal/Android-Issue-153521693/tree/cfb5033cf287572ac4b4972700f1656ce2b61d03/src/androidTest/java/de/skymatic/android_issue153521693

JaniruTEC committed:
Applied code style to BugTest.kt and SmallBugTest.kt

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Finalized work on BugTest.kt and SmallBugTest.kt

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added support for mapping cursors

Calling "requery" on a cursor may cause the driver to re-use queries, which is unwanted behavior.

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "Sql" to follow the API contract of "SupportSQLiteDatabase"

This also makes the class compatible with "MappingSupportSQLiteDatabase"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Moved logic for handling extra calls to dedicated class "OneOffDelegate"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Implemented "MappingSupportSQLiteQuery.bindTo" with "OneOffDelegate"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Applied fix for #529 [1] to re-enabled tests

See: b3449f07510fc86203e6b3ded5cfe3ff72902b01
Also See: c13b12ffc8d441f344793f2436506974b145899f

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Fixed "migrate12To14IndexSideEffects"

This also makes the test method compatible with "MappingSupportSQLiteDatabase"
Also added "assertIsUUID"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added "ValueExtractor" as argument type of "pseudo equality" methods

Added "ValueExtractor<T>" as type alias for "(T) -> Any?"
Changed argument declarations of "pseudo equality" methods and underlying implementations
Added "ValueExtractor<T>" as type declaration for existing value extractor sets

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Enforced single call convention for parts of "MappingSupportSQLiteQuery"

Also added workaround for tests

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "insert" and "compileStatement"

Changed "insert" to directly compile the mapped sql instead of creating a "MappingSupportSQLiteStatement"
Added check if the database is open to "compileStatement"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added test for "SupportSQLiteDatabase.compileStatement"

Also added missing "@VisibleForTesting" annotation to "MappingSupportSQLiteDatabase"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Improved caching to properly handle unconventional keys

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Moved database templating to separate dagger module

JaniruTEC committed:
Added handling for database corruption

If database corruption is detected the database is deleted and then recreated. Properly recreating and opening the new database involves invalidating any references to the old database and then running any initialization logic.

Added "Invalidatable" as wrapper for database references
Added "DatabaseOpenHelperFactory" to the dagger graph and added invalidation logic to "PatchedCallback"
Added dynamically delegating implementations for all entities to allow seamless swapping of underlying database instances
Configured Room to initialize based on the database template if a database with version 0 is encountered

JaniruTEC committed:
Added tests for verifying BOMs

"PatchedCallback" and "DatabaseOpenHelperFactoryKt.patchConfiguration" interact with the inner classes of "SupportSQLiteOpenHelper" in a way that is prone to introducing bugs due to unnoticed changes to those classes.
The tests introduced by this commit ensure that changes to the bills of materials for those classes are noticed and handled in the future.

JaniruTEC committed:
Made database templating more robust

Added additional validation
Moved database operations to "use" blocks

JaniruTEC committed:
Added "useFinally" to improve corruption handling

JaniruTEC committed:
Merge remote-tracking branch 'origin/develop' into feature/dao-migration

JaniruTEC committed:
Merge remote-tracking branch 'origin/develop' into feature/dao-migration

# Conflicts:
#	build.gradle

JaniruTEC committed:
Added test for opening previously corrupted/version 0 database

Also added "databaseName" to parameter list of "provideInternalCryptomatorDatabase" to allow different names for test databases

JaniruTEC committed:
Performed housekeeping

Added missing message in "CryptomatorAssert"
Fixed imports and codestyle
Added missing call to "disableWriteAheadLogging"
Made instance of "AOP_SQLiteDatabase" in "MappingSupportSQLiteDatabase" static

JaniruTEC committed:
Centralized database configuration

This creates a single source of truth for configuring the main database and any tests.

JaniruTEC committed:
Added logging for database corruption

JaniruTEC committed:
Added tests for verifying BOMs for dependencies of "MappingSupportSQLiteDatabase"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored tests for verifying BOMs

Moved tests from android tests to unit tests
Renamed containing class from "DatabaseOpenHelperFactoryTest" to "BomVerificationTest"
Fixed failing tests due to differences in JDK distros

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Improved tests for verifying BOMs

Added sorting of BOMs to ease debugging
Added check for classes that can't be verified by a unit test

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added tests for "MappingSupportSQLiteStatement.newBoundStatement"

Added inner class "MappingSupportSQLiteStatementTest" to "MappingSupportSQLiteDatabaseTest" and included an additional counter-based mapping
Added test "testNewBoundStatementSingle": Run "newBoundStatement" three times on the same "MappingSupportSQLiteStatement" (with the same unmapped sql strings and the same backing mapping)
Added test "testNewBoundStatementMultiple": Run "newBoundStatement" on three different instances of "MappingSupportSQLiteStatement" (with the same/different unmapped sql strings and per-statement backing mappings)
Added matching data sets for tests
Added a number of utility methods
Refactored "cartesianProduct" methods
Refactored extension methods for "ContentValues" defined in "MappingSupportSQLiteDatabaseTest.kt" and mirrored them for "Iterable<Any?>?"
Changed visibility of "MappingSupportSQLiteStatement.newBoundStatement" to "internal" and annotated it with "@VisibleForTesting"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added large tests for "MappingSupportSQLiteStatement.newBoundStatement"

Also added matching data sets
Added method "cartesianProductFour"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "testInsert" and tests in "MappingSupportSQLiteStatementTest"

Unified the implementation of "testInsert" with the implemention of all tests contained in "MappingSupportSQLiteStatementTest" by first augmenting and moving the implementation of "testSingleNewBoundStatement" to a new method "testSingleCompiledStatement" and then using that method in "testInsert":
> Previous control flows:
>> "testNewBoundStatementX" -> ("testConsecutiveNewBoundStatements" ->) "testConsecutiveNewBoundStatements" -> "testSingleNewBoundStatement" -> *Verifications*
>> "testInsert" -> *Verifications*
> New control flows:
>> "testNewBoundStatementX" -> ("testConsecutiveNewBoundStatements" ->) "testConsecutiveNewBoundStatements" -> "testSingleNewBoundStatement" -> "testSingleCompiledStatement" -> *Verifications*
>> "testInsert" -> "testSingleInsert" -> "testSingleCompiledStatement" -> *Verifications*

Removed extensions methods for "ContentValues" defined in "MappingSupportSQLiteDatabaseTest.kt" and replaced usages with calls to corresponding extension methods for "Iterable<Any?>?" defined in "MappingSupportSQLiteDatabaseTest.kt"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored "testInsertConflictAlgorithms" to also use "testSingleInsert"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Reformatted "MappingSupportSQLiteDatabaseTest.kt"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored calls to "runMigrationsAndValidate" in "UpgradeDatabaseTest"

Removed TODOs
Added delegating method with closing logic and default arguments

JaniruTEC committed:
Upgraded room to version 2.6.1

Added room-grade-plugin to build
Moved appropriate configuration to room-grade-plugin

JaniruTEC committed:
Refactored bindings in "MappingSupportSQLiteStatement"

Swapped bindings storage from lambdas to actual values
Added additional copy-operation to ByteArray-bindings to prevent downstream changes
Added utility methods

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Made "MappingSupportSQLiteStatement" threadsafe

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Performed cleanup

Added comments and documentation
Simplified implementation for "RandomUUIDMapping.mapWhereClause"
Replaced import alias "org.mockito.Mockito.`when` as whenCalled" with built-in "org.mockito.kotlin.whenever" in "MappingSupportSQLiteDatabaseTest.kt"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored initialization of "DatabaseOpenHelperFactory"

Moved call to "asCacheControlled" for singleton instance of "DatabaseOpenHelperFactory" from "DatabaseModule.provideInternalCryptomatorDatabase" to new binding "DatabaseModule.provideOpenHelperFactory" to prevent unnecessary duplicate invocations

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Refactored initialization of database template

Added logic to delete templates after use
Improved template verification
Changed binding type of templates from singleton to single-use
Changed type of exposed binding to "InputStream"
Updated "TemplateDatabaseContextTest", "CorruptedDatabaseTest", "CreateDatabaseTest" and "UpgradeDatabaseTest"

JaniruTEC committed:
Standardized database closing

Refactored "CreateDatabaseTest" and "UpgradeDatabaseTest" to close databases following the standard laid out by production logic/other tests

JaniruTEC committed:
Refactored WAL initialization

Moved WAL initialization from the "onConfigure" methods (via "applyDefaultConfiguration") of the database configurations to the corresponding instances of "SupportSQLiteOpenHelper"
Updated extension method "applyDefaultConfiguration" for "SupportSQLiteDatabase" defined in "CryptomatorDatabase.kt" to no longer configure WAL but verify that it has been set properly in the corresponding "SupportSQLiteOpenHelper"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Moved "pseudo equality" methods to dedicated class

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Moved helper methods for "Iterables" to dedicated class

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added type-checks to tests for "MappingSupportSQLiteQuery"

Added "@VisibleForTesting" to "MappingSupportSQLiteQuery" and changed its visibility to internal

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Standardized database configuration

Refactored database configuration in "CreateDatabaseTest" and "UpgradeDatabaseTest" to follow standard laid out by production logic/other tests

JaniruTEC committed:
Added missing WAL configuration/verification

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Prepared merging upstream v13

JaniruTEC committed:
Merge branch 'develop' into feature/dao-migration

JaniruTEC committed:
Finished merge of upstream v13 into production logic

Added missing fields defined by upstream v13 to "CloudEntity", "VaultEntity" and "AutoMigration14To15"
Added room-generated database schema for v15 as "15.json"
Moved "Upgrade12To13" and made it runnable
Replaced calls in "Upgrade12To13" with matching "pre15"-calls
Added "Upgrade12To13" to migration path in "DatabaseModule"

--

About this commit, as well as 4aa1c902 and c9a0d3db:

A new database version (13) introduced in the upstream by ac5f3ff2/#547 [1] (merged into develop by 7a37c033) needs to be merged into the "feature/dao-migration"-branch used for working on #506 [2].
4aa1c902 prepared the merge by bumping the versions defined by "feature/dao-migration" from 13 (l12-v1-r2-13) and 14 (l12-v2-r3-14) to 14 (l13-v1-r1-14) and 15 (l13-v2-r1-15).
l13-v1-r1-14 supersedes l12-v1-r2-13 and l13-v2-r1-15 supersedes l12-v2-r3-14 by applying the changes introduced by the upstream v13.
The relation between l13-v1-r1-14 and l13-v2-r1-15 is the same as between l12-v1-r2-13 and l12-v2-r3-14, i.e. the changes introduced by the upstream v13 are handled according to the semantics defined by that relation.
4aa1c902 deleted "13.json" (was l12-v1-r2-13) and replaced the content of "14.json" (was l12-v2-r3-14) with l13-v1-r1-14, which is the version that supersedes l12-v1-r2-13.
It therefore changes the migration path from [...] -> v12 -> v13 (l12-v1-r2-13) -> v14 (l12-v2-r3-14) to [...] -> v12 -> [Not yet merged upstream v13] -> v14 (l13-v1-r1-14) -> v15 (l13-v2-r1-15).

c9a0d3db merged the upstream v13 (c970438e) into 4aa1c902 and made the migration path continuous again: [...] -> v12 -> upstream v13 -> v14 (l13-v1-r1-14) -> v15 (l13-v2-r1-15)

This commit finishes this merge for the production source.
It adds "15.json" with l13-v2-r1-15 as its content, which is the version that supersedes l12-v2-r3-14.
Therefore "13.json" (l12-v1-r2-13) and "14.json" (l12-v2-r3-14) are superseded by "14.json" (l13-v1-r1-14) and "15.json" (l13-v2-r1-15) respectively.
It also applies l13-v2-r1-15 to "CloudEntity" and "VaultEntity", as well as "AutoMigration14To15".

[1] https://github.com/cryptomator/android/pull/547
[2] https://github.com/cryptomator/android/pull/506

Version definitions:
l12-v1-r2-13: e1bd29a5
l12-v2-r3-14: 23887b45
l13-v1-r1-14: 9536283c
l13-v2-r1-15: 074c77de

JaniruTEC committed:
Added "Upgrade12To13" to migration paths in tests

JaniruTEC committed:
Added "MigrationContainer"

Replaced raw arrays and consecutive invocations of migrations with usages of "MigrationContainer"

JaniruTEC committed:
Refactored "CorruptedDatabaseTest"

Moved initialization/property for "MigrationContainer" from "testOpenVersion0Database" to "setup"/class body
Cleaned up "testOpenVersion0Database"

JaniruTEC committed:
Added "testOpenVersion0DatabaseVerifyStreamAccessed"

JaniruTEC committed:
Added "testOpenDatabaseWithRecovery"

Replaced "openHelperFactory" property with method

JaniruTEC committed:
Made tests provided by upstream v13 runnable

Uncommented the tests
Removed outdated calls to "Upgrade0To1"
Replaced calls to removed method "applyTo" with calls to "migrate"

JaniruTEC committed:
Consolidated common test-logic to "DatabaseTests.kt"

Moved logic for configuring a test database, a "MigrationContainer" and creating a v0 database to "DatabaseTests.kt"

*This commit is related to issue #529 [1]*

[1] https://github.com/cryptomator/android/issues/529

JaniruTEC committed:
Added "assertOrder"

Added "assertOrder" to simplify orders that contain multiple calls to the same assertion

JaniruTEC committed:
Fixed ordered tests in "CorruptedDatabaseTest"

Fixed "testOpenVersion0DatabaseVerifyStreamAccessed":
"CorruptedDatabaseTest.kt:103" calls "RoomDatabase.compileStatement", which contains two calls to "writableDatabase" that were not previously accounted for.

Fixed "testOpenDatabaseWithRecovery":
The assertion in "CorruptedDatabaseTest.kt:129" is called by "PatchedCallback.onCreate" in "DatabaseOpenHelperFactory.kt:61", which is called by "FrameworkSQLiteOpenHelper.OpenHelper.innerGetDatabase". "PatchedCallback.onCreate" then calls "DatabaseCallback.onCreate", which then throws the "UnsupportedOperationException" ("DatabaseModule.kt:137") expected by the assertion in "CorruptedDatabaseTest.kt:123". However, it was not previously taken into account that "FrameworkSQLiteOpenHelper.OpenHelper.innerGetDatabase" discards the first thrown exception and tries again, which results in a second call to "PatchedCallback.onCreate" and therefore a second call to the assertion in "CorruptedDatabaseTest.kt:129" before the assertion in "CorruptedDatabaseTest.kt:123" is reached.
